### DISCUSSION TITLE
Move nodejs-admin-client to admin-ui repository

### DISCUSSION THREAD

**Discussion by stianst** [keycloak/keycloak#10122] [Ideas]
We have a rather awkward circular dependency between the following repositories:

* https://github.com/keycloak/keycloak
* https://github.com/keycloak/keycloak-nodejs-admin-client
* https://github.com/keycloak/keycloak-admin-ui

Let's take an example of a new feature that adds a new REST endpoint, as well as a new screen:

* Add REST API in keycloak
* Update admin client in keycloak-nodejs-admin-client
* Add screens in keycloak-admin-ui
* Include new UI in keycloak

Ideally of course from a dependency and release perspective would be to have everything into a single repository. However, since keycloak-admin-ui is a different team, and also uses completely different tools, that is not a direction we want to go.

However, we could simplify by at least merging keycloak-nodejs-admin-client into keycloak-admin-ui. That would at least eliminate one step, and since the keycloak-nodejs-admin-client and keycloak-admin-ui are maintained by the same team, that makes sense.

In the future we will also want to move account-ui, and likely also the client-side JavaScript adapter out of keycloak, which at the same time it would make life easier for us if there is a single repository for admin-ui, account-ui, client-side js, and admin node.js client.

**Comment by stianst** [comment]
@keycloak/keycloak-ui-maintainers FIY

**Comment by jonkoops** [comment]
I think this makes perfect sense, we have discussed this before as well in the team as something we might want to do. We can use NPM [workspaces](https://docs.npmjs.com/cli/v8/using-npm/workspaces) to link the packages together locally and develop them together.

It would also make sense to put the different UI projects together so that they can share dependencies between each other such as the build system and other common dependencies such as react. This would reduce the amount of effort we need to compile these projects and put them together.

I am very much in favor of this idea!

**Comment by ssilvert** [comment]
> Ideally of course from a dependency and release perspective would be to have everything into a single repository. However, since keycloak-admin-ui is a different team, and also uses completely different tools, that is not a direction we want to go.
> 
Definitely.  The UI team has benefitted tremendously from working in our own repository.  I could go on listing the ways, but it would be a very long list.  Also, as Keycloak has grown, it has become harder and harder to manage such a large repository with so many active developers.

The productivity gain from having our own repo was massive.

> However, we could simplify by at least merging keycloak-nodejs-admin-client into keycloak-admin-ui. That would at least eliminate one step, and since the keycloak-nodejs-admin-client and keycloak-admin-ui are maintained by the same team, that makes sense.
> 
I agree here.  This makes a lot of sense, especially because the Admin UI is the primary customer for the keycloak-nodejs-admin-client.

> In the future we will also want to move account-ui, and likely also the client-side JavaScript adapter out of keycloak, which at the same time it would make life easier for us if there is a single repository for admin-ui, account-ui, client-side js, and admin node.js client.

We have plenty of time to work this out in the future, but I don't see as much advantage here.   Those others don't share the same API, so a change to one is not likely to affect the other.  And there are bigger questions about the advantages of smaller repos versus the pain of integration.  We need to be careful with our choice.

Since at least three of us are in agreement about moving keycloak-nodejs-admin-client into keycloak-admin-ui, I'd like to see if @edewit agrees.  He has been the primary maintainer of keycloak-nodejs-admin-client.  So his opinion counts for a lot.

**Reply by stianst** [reply]
Putting the account-ui and js into the same repo would simplify things from a release perspective, as we can release a single project that goes into main Keycloak project, rather than releasing many independent projects.

End of the day it is mostly up to you guys, and we should talk about this in person. In general I'm leaning in the direction of making it your teams ownership to automate the releases of the components you own (admin-ui and nodejs-admin now, with account-ui and very likely frontend js client in the future). I have some ideas on how I want to drive the release, but I'm just going to initiate the release, which then your GA workflows should take action on.

**Reply by jonkoops** [reply]
I think it is perfectly possible to do this, but we'd have to invest in some monorepo tooling such as [TurboRepo](https://turborepo.org/) or [Nx](https://nx.dev/) so we can create a dependency graph between the different projects. This would allow us to make sure we are only testing projects that depend on each other, deducing time spent running tests and other CI tasks.

**Comment by edewit** [comment]
I'm :+1: on this and I think it makes a lot of sense to bundle it, but I don't see how it would solve the problem that you describe. It would "only" remove one step in the dependency chain, a change in the rest api would still mean updating "keycloak" twice.

**Reply by stianst** [reply]
Not a complete solution for sure, but would be a simplification.

**Comment by ssilvert** [comment]
@stianst Since we are all in agreement about moving keycloak-nodejs-admin-client into keycloak-admin-ui, the only question now is when? 

I don't really have an opinion on how soon this should happen.  Of course it takes time from our other work, but if you have any reason why it should happen sooner or later I'd like to know.  Otherwise, I'll just work with the team to try to figure out the smoothest time to make it happen.

**Reply by stianst** [reply]
@ssilvert nothing urgent from my side. Just give me a heads up when you do it as I would need to adjust the Keycloak release jobs.

I am thinking about how to redoing the release process, but that's a fairly long term thing, not something I want to do right now. Would like to chat to you about this at some point though.
